torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -572,9 +574,39 @@ generateIndexAndExtentMap(
     auto c_tv = c2p_tv_stack[i];
     auto p_tv = c2p_tv_stack[i + 1];
 
-    // Map root ID's from consumer to producer
-    auto c2p_root_map =
-        TensorDomain::mapRootCtoP(c_tv->domain(), p_tv->domain());
+    // Map root ID's from consumer to producer. c2p_tv_stack may have
+    // an additional producer tensor that is fully replayed. It may
+    // not be actually computed at the consumer. It needs to be
+    // processed specially as it needs full mapping even when it could
+    // indicate invalid root mapping in the sense of computeAt
+    // viability. For the particular case, the simpler pairwise
+    // mapping just works as they are guaranteed to be a
+    // producer-consumer pair.
+    std::unordered_map<IterDomain*, IterDomain*> c2p_root_map;
```

Comment:
Sorry I'm probably not up to date on this. I was wondering where the additional `producer` comes from, and it seems to be the second last one on the `c2p stack`. I knew that the multiple `fusion` outputs are forced to compute at one another. Is it a similar scenario but for an additional `producer` here?

Reply:
It's the producer tensor of `getProducerIndex` or `getGlobalProducerIndex`. We create a list of TVs by following the `computeAtView` field of the consumer tensor. We reverse it and append the producer to it. 

Also, though the list is named as `c2p`, it's the computeAt relationship that actually defines it.

Reply:
Oh I see! Thanks for the clarification. Just wanted to double check if I'm understanding correctly. This case here is treated differently from the remaining of the `tv stack` because `ComputeAtRootDomainMap` may not want to include their root maps due to potential inconsistency or reduction violations, right?

Reply:
Yes. The computeAt in our system defines not just where a tensor is defined but also where it is declared (allocated). When that tensor is used by multiple consumers, we need to make sure it is accessible by all its consumers. That's the logic behind the validation done in `ComputeAtRootDomainMap`. 

The tensors in the computeAt stack are the ones that are transformed based on the mapping provided by `ComputeAtRootDomainMap`. So, at this point, what we do is to transform index expressions by traversing the computeAt stack. We transform indices defined for one tensor to those for its next next based on the root mapping. 

In the special case with the additional producer tensor, the producer may not be computed at the consumer, and the only thing we can say here is that it's a producer of the consumer. So, `ComputeAtRootDomainMap` may return no mapping for this producer-consumer pair. Instead of `ComputeAtRootDomainMap`, `PairwiseRootDomainMap` simply looks at a producer-consumer pair and maps each axis. Though it's only valid for producer-consumer pairs, it doesn't care the computeAt semantics, and that's why it is used for the special case.

Note that `PairwiseRootDomainMap` may not work for the tensors originally in the computeAt stack since computeAt does not necessarily mean a producer-consumer relationship, i.e., terminating output tensors may have computeAt relationships, but by definition they are not producer-consumer. So, `ComputeAtRootDomainMap` is used as it can be used with arbitrary pairs of tensors.

All in all, in `getProducerIndex`, `PairwiseRootDomainMap` is used for the producer-consumer arguments. After that, `ComputeAtRootDomainMap` is used for the "real" computeAt tensors traversed from the consumer.

Reply:
This is very helpful - can we capture some of it as code comments?

Reply:
Will do.

Reply:
Added. bad3543